#### Use the make_regression() function to create a synthetic binary classification problem with 1,000 examples and 20 input features. Use this synthetic dataset to build a classification model using Random forest regressor. Evaluate your model using KFold validation.

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.model_selection import cross_val_score, KFold
from sklearn.datasets import make_regression
from sklearn.decomposition import PCA
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score

In [2]:
X, y = make_regression(n_samples=1000, n_features=20, random_state=42)

In [3]:
X_train, X_test, y_train, y_test  = train_test_split(X, y, test_size=0.2,
                                                     random_state=42)

In [4]:
rfr = RandomForestRegressor(random_state=42, n_jobs=-1)
params = {'n_estimators':[100, 400, 800 ],'max_features': [15,20]}

gs = GridSearchCV(estimator=rfr, param_grid=params, scoring='r2', cv=5)

In [5]:
gs.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=RandomForestRegressor(n_jobs=-1, random_state=42),
             param_grid={'max_features': [15, 20],
                         'n_estimators': [100, 400, 800]},
             scoring='r2')

In [6]:
print(gs.best_params_)
print(gs.best_estimator_)
print(gs.best_score_)

{'max_features': 15, 'n_estimators': 800}
RandomForestRegressor(max_features=15, n_estimators=800, n_jobs=-1,
                      random_state=42)
0.7805412910049704


In [8]:
kf = KFold(n_splits=4, shuffle=True, random_state=42)

r2_kf = cross_val_score(gs, X_train, y_train, cv=kf, n_jobs=-1, scoring='r2')
print('KFold on Train set')
print('R2 Score {:.4f}'.format(r2_kf.mean()))

KFold on Train set
R2 Score 0.7845


In [9]:
print('R2 on Test set')
print('R2 Score {:.4f}'.format(r2_score(y_test, gs.predict(X_test))))

R2 on Test set
R2 Score 0.7718


### Observing R-sqaure as 0.7845 and 0.7718 on KFold(train set) and Test set respectively.